viewer/src/utilities/IndexSet.ts

```diff
@@ -0,0 +1,594 @@
+        return newRight.addRange(unioned);
+      } else if (newRight == undefined && newLeft != undefined) {
+        // ---"---
+        return newLeft.addRange(unioned);
+      }
+
+      // "as" terms to avoid compiler complaints
+      const newNode = IntermediateIndexNode.fromIndexNodesAndBalance(newLeft as IndexNode, newRight as IndexNode);
+
+      return newNode.addRange(unioned);
+    } else if (canUnionLeft) {
+      return IntermediateIndexNode.fromIndexNodesAndBalance(this.left.addRange(range), this.right);
+    } else if (canUnionRight) {
+      return IntermediateIndexNode.fromIndexNodesAndBalance(this.left, this.right.addRange(range));
+    } else {
+      // Range lies between ranges of left and right subtree,
+      // add to smallest
+      if (this.left.maxSubtreeDepth < this.right.maxSubtreeDepth) {
+        return IntermediateIndexNode.fromIndexNodesAndBalance(this.left.addRange(range), this.right);
+      } else {
+        return IntermediateIndexNode.fromIndexNodesAndBalance(this.left, this.right.addRange(range));
+      }
+    }
+  }
+
+  removeRange(range: NumericRange): IndexNode | undefined {
+    // If input range does not intersect with this range, return
+    if (!range.intersects(this.range)) {
+      return this;
+    }
+
+    const [newThis, soakedRange] = this.soak(range);
+
+    let leftRange: NumericRange | undefined = undefined;
+    let rightRange: NumericRange | undefined = undefined;
+
+    // If soakedRange extends to either the left or right to the
+    // numeric range, we take the ranges extending out and insert anew.
+    if (soakedRange.from < range.from) {
+      leftRange = NumericRange.createFromInterval(soakedRange.from, range.from - 1);
+    }
+
+    if (soakedRange.toInclusive > range.toInclusive) {
+      rightRange = NumericRange.createFromInterval(range.toInclusive + 1, soakedRange.toInclusive);
+    }
+
+    if (newThis == undefined) {
+      // If all ranges in this was soaked up, create new node with
+      // non-empty left and right ranges
+      if (leftRange != undefined && rightRange != undefined) {
+        return IntermediateIndexNode.fromIndexNodesAndBalance(
+          new LeafIndexNode(leftRange),
+          new LeafIndexNode(rightRange)
+        );
+      } else if (leftRange != undefined) {
+        return new LeafIndexNode(leftRange);
+      } else if (rightRange != undefined) {
+        return new LeafIndexNode(rightRange);
+      } else {
+        return undefined;
+      }
+    } else {
+      // Add non-empty left- and right ranges
+
+      let nodeToReturn = newThis;
+
+      if (leftRange != undefined) {
+        nodeToReturn = nodeToReturn.addRange(leftRange);
+      }
+
+      if (rightRange != undefined) {
+        nodeToReturn = nodeToReturn.addRange(rightRange);
+      }
+
+      return nodeToReturn;
+    }
+  }
+
+  balance(): IntermediateIndexNode {
+    const d0 = this.left.maxSubtreeDepth;
+    const d1 = this.right.maxSubtreeDepth;
+
+    if (d1 + 2 <= d0) {
```

Comment:
define what 2 is, and use `const leftSubTreeSize`, and `const rightSubTreeSize`, d0 and d1 doesn't tell me anything. We can afford the extra bytes 😉